### ISSUE TITLE
Implementation of DCLW

### ISSUE THREAD

**Issue by wqtwjt1996** [raminnakhli/Decoupled-Contrastive-Learning#10]
Hi,

I saw your implementation of the nominator of DCLW weight_fn, which uses element multiplication for the z1 and z2 calculation.
But in the paper, the nominator of DCLW weight_fn formula is: exp(<z1, z2> / sigma). Can you tell me why you use element multiplication for the z1 and z2 calculation instead of torch.mm(z1, z2) or dot multiplication? 

Thanks.

**Comment by raminnakhli** [comment]
Hi @wqtwjt1996, thanks for the interest and question. I believe this is the corresponding part in the paper:

As you can see, the nominator includes multiplication of each element in z1 into the corresponding element of z2.  Please let me know if this doesn't make sense. Otherwise, I'll close the ticket.
<img width="460" alt="dclw" src="https://user-images.githubusercontent.com/35274390/197066741-870cf043-7b79-4fe3-800f-ab8c05531e98.png">

**Comment by wqtwjt1996** [comment]
Hi @raminnakhli Thanks for the reply. 

![](https://user-images.githubusercontent.com/35274390/197066741-870cf043-7b79-4fe3-800f-ab8c05531e98.png)
<img width="629" alt="截屏2022-10-27 下午2 30 34" src="https://user-images.githubusercontent.com/22970429/198370482-13e778f7-9115-48ad-845e-fd8adcb4b213.png">
 As you can see from the two pictures, the formulas are similar which all include: exp(<z_1, z_2>). But in your code, for formula(5) (6) you use matrix multiplication, and for the formula of w(z_1, z_2), you use element multiplication.

Could you please explain to me why? Thanks!

**Comment by raminnakhli** [comment]
Hi @wqtwjt1996,

The matrix multiplication is for the calculation of the U variable that you can see in the denominator of the equation (5). As you can see, in line 27 of the loss file, I'm using the `diag` to get the diagonal elements which is equivalent to performing element-wise multiplication. 

`positive_loss = -torch.diag(cross_view_distance) / self.temperature`

I believe this would answer the question. Therefore, I'd close the ticket. Feel free to follow up.
